Std/Data/Array/Lemmas.lean

```diff
@@ -724,3 +724,182 @@ theorem all_eq_true_iff_forall_mem {l : Array α} : l.all p ↔ ∀ x, x ∈ l 
 /-! ### erase -/
 
 @[simp] proof_wanted erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a
+
+/-! ### swap -/
+
+@[simp] theorem swap_def_jeqi {a: Array α} {i j : Nat}
+    (_ : i < a.size := by trivial) (_ : j < a.size := by trivial)
+    (_ : j < (a.swap ⟨i, ‹_›⟩ ⟨j, ‹_›⟩).size := by trivial) : (a.swap ⟨i, _⟩ ⟨j, _⟩)[j] = a[i] :=
+  let i : Fin a.size := ⟨i, _⟩
+  let j : Fin a.size := ⟨j, _⟩
+
+  show (a.swap i j)[j] = a[i] from aux
+where aux {i j : Fin a.size} : (a.swap i j)[j] = a[i] := by simp [swap, fin_cast_val]
+
+@[simp] theorem swap_def_ieqj {a: Array α} {i j : Nat}
+    (_ : i < a.size := by trivial) (_ : j < a.size := by trivial)
+    (_ : i < (a.swap ⟨i, ‹_›⟩ ⟨j, ‹_›⟩).size := by trivial) : (a.swap ⟨i, _⟩ ⟨j, _⟩)[i] = a[j] :=
+  let i : Fin a.size := ⟨i, _⟩
+  let j : Fin a.size := ⟨j, _⟩
+
+  show (a.swap i j)[i] = a[j] from aux
+where aux {i j : Fin a.size} : (a.swap i j)[i] = a[j] :=
+  let vi := a.get i
+  let vj := a.get j
+  let j := a.size_set i vj ▸ j
+
+  let a := a.set i vj
+  let a' := a.set j vi
+
+  suffices a'[i.val]'(by simp [a, a']) = vj from by simp [a'] at this; assumption
+  if he : j.val = i.val then
+    by simp [←he, vi, vj, j, a', fin_cast_val]
+  else
+    suffices a[i.val]'(by simp [a]) = vj from (a.get_set_ne j vi _ he).symm ▸ this
+    by simp [get_set_ne, a]
+
+@[simp] theorem swap_def_else {a: Array α} {i j : Fin a.size}
+    (_ : p < a.size := by trivial) (_ : p < (a.swap i j).size := by trivial)
+    (_ : i.val ≠ p) (_ : j.val ≠ p) : (a.swap i j)[p] = a[p] :=
+  let i : Fin a.size := ⟨i, _⟩
```

Comment:
Does this line have any effect?

I haven't read your proofs, but they seem longer than necessary.

Reply:
sorry about this good catch, let me look over the proofs more carefully .. I wrote them a long time ago. Will try and shorten them.